test/integration_test.cpp

```
@@ -358,18 +382,87 @@ b = measure qq[0]
     }
 }
 
+TEST_F(IntegrationTest, correct_init_is_just_discarded_by_the_simulator) {
+    auto program = R"(
+version 3.0
+
+qubit[2] q
+
+init q[0]
+H q[0]
+barrier q[1]
+init q[1]
+CNOT q[0], q[1]
+)";
+    std::size_t iterations = 1;
+    auto actual = run_from_string(program, iterations, "3.0");
+
+    // Expected 'q' state should be |00>+|10>
+    // State is |00>+|11> after creating the Bell state
+    // A correct 'init' instruction is discarded by the simulator
+    EXPECT_EQ(actual.state,
+        (SimulationResult::State{
+            { "00", core::Complex{ .real = 1 / std::sqrt(2), .imag = 0, .norm = 0.5 } },
+            { "11", core::Complex{ .real = 1 / std::sqrt(2), .imag = 0, .norm = 0.5 } }
+    }));
+}
+
+TEST_F(IntegrationTest, incorrect_init_throws_error) {
+    auto program = R"(
+version 3.0
+
+qubit[2] qq
+
+H qq[0]
+CNOT qq[0], qq[1]
+init qq[0]
+)";
+    auto result = execute_string(program);
+    auto message = std::string{ std::get_if<SimulationError>(&result)->what() };
+    EXPECT_TRUE(std::holds_alternative<SimulationError>(result));
+    EXPECT_EQ(message, "trying to 'init qq[0]' but qubit is not in ground state");
+}
+
+TEST_F(IntegrationTest, barrier_and_wait_are_just_discarded_by_the_simulator) {
+    auto program = R"(
+version 3.0
+
+qubit[2] q
+
+barrier q
+wait(0) q[0]
+H q[0]
+barrier q[0]
+wait(-1) q[1]
```

Shouldn't libQASM throw an error here, i.e., with `wait(-1)`?

---

Yes and no.

Yes because, in this case, the time parameter is a constant value, so it may be dealt with at parsing time.

But generally no for the ever recurring reason, the same as to why we would have problems checking that `init` is performed on a 'dirty' qubit: the time parameter for `wait` or the index operand for `init` could be variable vales, only known at runtime.

Simulation is kind of interpretation, you execute every instruction alongside your system is running, so by the time you are about to execute `wait` or `init` you already know the time parameter and the index operand.

Here, libQASM only checks that the time parameter is of an integer type. Should we throw an exception in the simulator when we receive a negative time parameter? Something like "incorrect 'wait({}) {}[{}]': time cannot be negative"? If so, I'll add that check and a test for it.